test/pjrt/test_runtime_single_proc_gpu.py

```diff
@@ -38,12 +38,35 @@ def test_num_local_devices(self):
     self.assertLen(xm.get_xla_supported_devices(),
                    xr.addressable_device_count())
     self.assertEqual(self.num_cuda_devices, xr.addressable_device_count())
+    self.assertEqual(self.num_cuda_devices, xr.local_device_count())
 
-  def test_num_global_devices(self):
-    self.assertLen(torch_xla._XLAC._xla_get_all_devices(),
-                   xr.global_device_count())
+  def test_num_global_device_count(self):
     self.assertEqual(self.num_cuda_devices, xr.global_device_count())
 
+  def test_local_process_count(self):
+    self.assertEqual(1, xr.local_process_count())
+
+  # xw32: need review here.
+  def test_world_size(self):
```

Comment:
hey @will-cromar , on the single-processing case, is 1 the expected output of xr.world_size()?

Reply:
Either case, it's the number of logical devices. For SPMD, that's one per process, and normally one per host. For multiprocessing, this is the number of global devices.

Think of this as the number of `torch.device`s in the whole system. SPMD bundles all runtime devices as one `torch.device`, and multiprocessing exposes each one individually.

Reply:
Thanks for the explanation. I get confused when reading PyTorch's doc https://pytorch.org/docs/stable/distributed.html which defines `world_size` as `Number of processes participating in the job.`. I really hope both PyTorch and PyTorch/XLA can unify the definition.

Does the expectation below look correct to you?
- Here, in the non-SPMD single-processing test, `xr.world_size()` should return 1.
- In the non-SPMD multi-processing test (the test `test_world_size` in pytorch/xla/test/pjrt/test_runtime_multi_gpu.py in this PR), `xr.world_size()` should return the number of GPU devices.
- In the SPMD case, `xr.world_size()` should return 1.

Reply:
> I really hope both PyTorch and PyTorch/XLA can unify the definition.

If you except our multithreading hackery for old TPUs (or treat threads as processes for this definition), these definitions do align most of the time. Each process/thread should get one logical device if you start through `xmp.spawn` or `torchrun`.

> Here, in the non-SPMD single-processing test, xr.world_size() should return 1.

I don't think that's true after your fix to expose all available GPUs to a single process by default. You would see devices `['xla:0', 'xla:1', 'xla:2', 'xla:3']`. Unless you don't set `xm.set_replication_devices` --  then `xm.world_size` is 1... I want to simplify this latter condition out, but we need either refactoring or an explicit API change.

Reply:
I see. I also spent some time digging the issue and here is what I found:

1. The only place where we use `xr.world_size()` is [here](https://github.com/pytorch/xla/blob/1d4a8da5f8e0507731ef2178aa7d24495faf9bf4/torch_xla/_internal/rendezvous.py#L48) and it's the place where we interact with upstream PyTorch: if it's spmd, we assign `xr.process_count()` to `world_size`; otherwise, we assign `xr.world_size()`. If the `world_size` there will be used by pytorch, then I feel it should follow pytorch's definition of `world_size` ( the number of processes instead of number of devices). What do you think?

2. On TPU, the non-SPMD single-processing, I can see:
```
# PJRT_DEVICE=TPU python
>>> xr.global_device_count()
8
>>> xr.world_size()
1
>>> xr.local_device_count()
8
>>> xr.addressable_device_count()
8
```

So I feel the discussion boils down to how we define world_size, whether it's the number of devices or the number of processes. If it's the latter, both TPU and GPU are already doing the right thing. Otherwise, we may need to change both.

Reply:
discussed offline, will assert equal to 1